File: raffia-core/src/main/java/com/codepoetics/raffia/builders/BasketWeaver.java
package com.codepoetics.raffia.builders;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.injections.Injections;
import com.codepoetics.raffia.mappers.Mapper;
import com.codepoetics.raffia.mappers.Mappers;
import com.codepoetics.raffia.visitors.Visitors;
import com.codepoetics.raffia.writers.BasketWeavingWriter;
import com.codepoetics.raffia.writers.BasketWriter;
import com.codepoetics.raffia.writers.Writers;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public final class BasketWeaver implements BasketWriter<BasketWeaver> {

  public static BasketWeaver create() {
    return weavingWith(Writers.weaving());
  }

  public static BasketWeaver weavingWith(BasketWeavingWriter writer) {
    return new BasketWeaver(writer);
  }

  private final BasketWeavingWriter writer;

  private BasketWeaver(BasketWeavingWriter writer) {
    this.writer = writer;
  }

  public Basket weave() {
    return writer.weave();
  }

  private BasketWeaver with(BasketWeavingWriter writer) {
    return new BasketWeaver(writer);
  }

  @Override
  public BasketWeaver beginObject() {
    return with(writer.beginObject());
  }

  @Override
  public BasketWeaver beginArray() {
    return with(writer.beginArray());
  }

  @Override
  public BasketWeaver end() {
    return with(writer.end());
  }

  @Override
  public BasketWeaver key(String key) {
    return with(writer.key(key));
  }

  @Override
  public BasketWeaver add(String value) {
    return with(writer.add(value));
  }

  @Override
  public BasketWeaver add(BigDecimal value) {
    return with(writer.add(value));
  }

  @Override
  public BasketWeaver add(boolean value) {
    return with(writer.add(value));
  }

  @Override
  public BasketWeaver addNull() {
    return with(writer.addNull());
  }

  public BasketWeaver add(String key, String value) {
    return key(key).add(value);
  }

  public BasketWeaver add(String key, BigDecimal value) {
    return key(key).add(value);
  }

  public BasketWeaver add(String key, boolean value) {
    return key(key).add(value);
  }

  public BasketWeaver add(String key, Basket value) {
    return key(key).add(value);
  }

  public BasketWeaver array() {
    return with(writer.beginArray().end());
  }

  public BasketWeaver array(Basket firstItem, Basket...subsequentItems) {
    return array(firstItem, subsequentItems, Mappers.<Basket>id());
  }

  public BasketWeaver array(String firstItem, String...subsequentItems) {
    return array(firstItem, subsequentItems, Injections.fromString);
  }

  public BasketWeaver array(BigDecimal firstItem, BigDecimal...subsequentItems) {
    return array(firstItem, subsequentItems, Injections.fromNumber);
  }

  public BasketWeaver array(boolean firstItem, Boolean...subsequentItems) {
    return array(firstItem, subsequentItems, Injections.fromBoolean);
  }

  private <T> BasketWeaver array(T first, T[] subsequent, Mapper<T, Basket> mapper) {
    BasketWeavingWriter state = writer.beginArray();
    state = state.add(mapper.map(first));
    for (T item : subsequent) {
      state = state.add(mapper.map(item));
    }
    return with(state.end());
  }

  public BasketWeaver array(Collection<Basket> items) {
    BasketWeavingWriter state = writer.beginArray();
    for (Basket basket : items) {
      state = state.add(basket);
    }
    return with(state.end());
  }

  public BasketWeaver addArray(String key) {
    return key(key).array();
  }

  public BasketWeaver addArray(String key, Basket firstItem, Basket...subsequent) {
    return key(key).array(firstItem, subsequent);
  }

  public BasketWeaver addArray(String key, String firstItem, String...subsequent) {
    return key(key).array(firstItem, subsequent);
  }

  public BasketWeaver addArray(String key, BigDecimal firstItem, BigDecimal...subsequent) {
    return key(key).array(firstItem, subsequent);
  }

  public BasketWeaver addArray(String key, boolean firstItem, Boolean...subsequent) {
    return key(key).array(firstItem, subsequent);
  }

  public BasketWeaver add(Basket value) {
    return with(writer.add(value));
  }

  @Override
  public String toString() {
    return "Basket weaver with state: " + writer;
  }

}

File: raffia-core/src/main/java/com/codepoetics/raffia/builders/BasketWeaver.kt
package com.codepoetics.raffia.builders

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.writers.BasketWeavingWriter
import com.codepoetics.raffia.writers.BasketWriter
import com.codepoetics.raffia.writers.Writers

import java.math.BigDecimal

class BasketWeaver private constructor(private val writer: BasketWeavingWriter) : BasketWriter<BasketWeaver> {

    fun weave(): Basket = writer.weave()

    private fun with(writer: BasketWeavingWriter): BasketWeaver = BasketWeaver(writer)

    override fun beginObject(): BasketWeaver = with(writer.beginObject())

    override fun beginArray(): BasketWeaver = with(writer.beginArray())

    override fun end(): BasketWeaver = with(writer.end())

    override fun key(key: String): BasketWeaver = with(writer.key(key))

    override fun add(value: String): BasketWeaver = with(writer.add(value))

    override fun add(value: BigDecimal): BasketWeaver = with(writer.add(value))

    override fun add(value: Boolean): BasketWeaver = with(writer.add(value))

    override fun addNull(): BasketWeaver = with(writer.addNull())

    fun add(key: String, value: String): BasketWeaver = key(key).add(value)

    fun add(key: String, value: BigDecimal): BasketWeaver = key(key).add(value)

    fun add(key: String, value: Boolean): BasketWeaver = key(key).add(value)

    fun add(key: String, value: Basket): BasketWeaver = key(key).add(value)

    fun array(): BasketWeaver = with(writer.beginArray().end())

    fun array(firstItem: Basket, vararg subsequentItems: Basket): BasketWeaver =
            array(firstItem, subsequentItems) { it }

    fun array(firstItem: String, vararg subsequentItems: String): BasketWeaver =
            array<String>(firstItem, subsequentItems) { Basket.ofString(it) }

    fun array(firstItem: BigDecimal, vararg subsequentItems: BigDecimal): BasketWeaver =
            array<BigDecimal>(firstItem, subsequentItems) { Basket.ofNumber(it) }

    fun array(firstItem: Boolean, vararg subsequentItems: Boolean): BasketWeaver =
            array<Boolean>(firstItem, subsequentItems.toTypedArray()) { Basket.ofBoolean(it) }

    private fun <T> array(first: T, subsequent: Array<out T>, mapper: (T) -> Basket): BasketWeaver =
        with(subsequent.fold(
                writer.beginArray().add(mapper(first)),
                { state, item -> state.add(mapper(item)) }).end())

    fun array(items: Collection<Basket>): BasketWeaver =
        with(items.fold(writer.beginArray()) { state, item -> state.add(item) }.end())

    fun addArray(key: String): BasketWeaver = key(key).array()

    fun addArray(key: String, firstItem: Basket, vararg subsequent: Basket): BasketWeaver =
            key(key).array(firstItem, *subsequent)

    fun addArray(key: String, firstItem: String, vararg subsequent: String): BasketWeaver =
            key(key).array(firstItem, *subsequent)

    fun addArray(key: String, firstItem: BigDecimal, vararg subsequent: BigDecimal): BasketWeaver =
            key(key).array(firstItem, *subsequent)

    fun addArray(key: String, firstItem: Boolean, vararg subsequent: Boolean): BasketWeaver =
            key(key).array(firstItem, *subsequent)

    fun add(value: Basket): BasketWeaver {
        return with(writer.add(value))
    }

    override fun toString(): String = "Basket weaver with state: " + writer

    companion object {

        @JvmStatic
        fun create(): BasketWeaver {
            return weavingWith(Writers.weaving())
        }

        @JvmStatic
        fun weavingWith(writer: BasketWeavingWriter): BasketWeaver {
            return BasketWeaver(writer)
        }
    }

}
